File: src/main/java/ch20_generics/Mixins.cpp
/**
 * @author runningpig66
 * @date 2026/1/15 周四
 * @time 1:00
 * P.725 §20.14 混型 §20.14.1 C++ 中的混型
 *
 * 在 C++ 中，使用多重继承的最重要原因之一就是混型。实现混型的一种更优雅的方式是使用参数化类型，
 * 因为混型就是继承自其类型参数的类。在 C++ 中，你可以轻松地混型，因为 C++ 能记住其模板参数的类型。
 * 下面是一个 C++ 的示例，带有两个混型类型。其中一个给每个对象实例都混入了自带时间戳这样一种属性，另一个则给每个对象实例混入了一个序列号：
 */
#include <string>
#include <ctime>
#include <iostream>
using namespace std;

template<class T> class TimeStamped : public T {
  long timeStamp;
public:
  TimeStamped() { timeStamp = time(0); }
  long getStamp() { return timeStamp; }
};

template<class T> class SerialNumbered : public T {
  long serialNumber;
  static long counter;
public:
  SerialNumbered() { serialNumber = counter++; }
  long getSerialNumber() { return serialNumber; }
};

// Define and initialize the static storage:
template<class T> long SerialNumbered<T>::counter = 1;

class Basic {
  string value;
public:
  void set(string val) { value = val; }
  string get() { return value; }
};

// 在 main() 方法中，mixin1 和 mixin2 所产生的类型都拥有所混入的类型的所有方法。你可以将混型想象为一种将已有的类映射到新的子类上的功能。
// 可以看到，用这种技术创建混型是多么简单。基本上，你只需要声明“这就是我想要的”，然后它就出现了：
int main() {
  TimeStamped<SerialNumbered<Basic>> mixin1, mixin2;
  mixin1.set("test string 1");
  mixin2.set("test string 2");
  cout << mixin1.get() << " " << mixin1.getStamp() <<
    " " << mixin1.getSerialNumber() << endl;
  cout << mixin2.get() << " " << mixin2.getStamp() <<
    " " << mixin2.getSerialNumber() << endl;
}
/* Output:
test string 1 1452987605 1
test string 2 1452987605 2
*/

File: src/main/java/ch20_generics/Mixins.java
package ch20_generics;

import java.util.Date;

/**
 * @author runningpig66
 * @date 2026/1/15 周四
 * @time 1:03
 * P.726 §20.14 混型 §20.14.2 与接口混合
 * <p>
 * 一种常见的推荐方案是使用接口来达到混型的效果，就像这样：
 * <p>
 * Mixin 类基本上是使用了委托机制（delegation），因此每个被混入的类型都需要在 Mixin 中有一个字段，
 * 而你必须在 Mixin 中编写所有必要的方法来将调用转发到合适的对象上。本例中的类都很简单，如果是更复杂的混型，则代码量会大幅增长。
 */
interface TimeStamped {
    long getStamp();
}

class TimeStampedImp implements TimeStamped {
    private final long timeStamp;

    TimeStampedImp() {
        timeStamp = new Date().getTime();
    }

    @Override
    public long getStamp() {
        return timeStamp;
    }
}

interface SerialNumbered {
    long getSerialNumber();
}

class SerialNumberedImp implements SerialNumbered {
    private static long counter = 1;
    private final long serialNumber = counter++;

    @Override
    public long getSerialNumber() {
        return serialNumber;
    }
}

interface Basic {
    void set(String val);

    String get();
}

class BasicImp implements Basic {
    private String value;

    @Override
    public void set(String val) {
        value = val;
    }

    @Override
    public String get() {
        return value;
    }
}

// 这个例子本质上是委托（Delegation）。这就是手动实现的委托，目的是为了模拟多重继承。
class Mixin extends BasicImp implements TimeStamped, SerialNumbered {
    private TimeStamped timeStamp = new TimeStampedImp();
    private SerialNumbered serialNumber = new SerialNumberedImp();

    @Override
    public long getStamp() {
        return timeStamp.getStamp();
    }

    @Override
    public long getSerialNumber() {
        return serialNumber.getSerialNumber();
    }
}

public class Mixins {
    public static void main(String[] args) {
        Mixin mixin1 = new Mixin(), mixin2 = new Mixin();
        mixin1.set("test string 1");
        mixin2.set("test string 2");
        System.out.println(mixin1.get() + " " + mixin1.getStamp() + " " + mixin1.getSerialNumber());
        System.out.println(mixin2.get() + " " + mixin2.getStamp() + " " + mixin2.getSerialNumber());
    }
}
/* Output:
test string 1 1768411262991 1
test string 2 1768411262991 2
 */
